### ISSUE TITLE
Gradients not computed upon `tf.concat()`

### ISSUE THREAD

**Issue by joonjeon** [tensorflow/tensorflow#55648]
Hi!

I am trying to re-write MAVNet (https://github.com/sudakshin/imitation_learning/blob/master/2.train_model/MavNet.py) into genuine Tensorflow (instead of TFLearn) and train it. Note that MAVNet consists of numerous `tf.keras.layers.Concatenate()`, which involves `tf.concat()`.
For some reason, gradient flow seems to break... and I just found out that the flow breaks whenever it passes `tf.concat()`.

Please pardon if this issue is a duplicate of #37726 , but I feel that the previous issue has been closed too early without being thoroughly investigated and resolved.

The details of this issue is as follows:

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.3.0-2.8.0 (issue reproduced from all versions included in this range)
- Python version: 3.8.3
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: CUDA 11.4 + cuDNN 8.1.1
- GPU model and memory: NVIDIA GeForce GTX Titan X

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
Tensorflow is not able to compute gradients after concatenating multiple feature maps with `tf.concat()`.

**Describe the expected behavior**
Tensorflow should be able to compute gradients after concatenating multiple feature maps with `tf.concat()`.

**Standalone code to reproduce the issue**

(Here, we assume all necessary libraries (including Tensorflow) are imported)

    class LocalResponseNormalization(tf.keras.layers.Layer):
        def __init__(self, depth_radius=5, bias=1, alpha=1, beta=0.5):
            super(LocalResponseNormalization, self).__init__()
            self.depth_radius = depth_radius
            self.bias = bias
            self.alpha = alpha
            self.beta = beta

        def build(self,input):
            pass

        def call(self, input):
            return tf.nn.local_response_normalization(input, alpha=self.alpha, beta=self.beta)

        def get_config(self):
            config = super().get_config().copy()
            return config

    def MavNet(input_shape, num_classes=2):
        img_input = tf.keras.layers.Input(shape=input_shape)

        conv1_7_7 = tf.keras.layers.Conv2D(64, 7, strides=2, activation='relu', padding='same', name = 'conv1_7_7_s2')(img_input)
        pool1_3_3 = tf.keras.layers.MaxPooling2D(pool_size=3, strides=2)(conv1_7_7)
        pool1_3_3 = LocalResponseNormalization(alpha=0.0001, beta=0.75)(pool1_3_3)
        conv2_3_3_reduce = tf.keras.layers.Conv2D(64,1, activation='relu', padding='same', name = 'conv2_3_3_reduce')(pool1_3_3)
        conv2_3_3 = tf.keras.layers.Conv2D(192,3, activation='relu', padding='same', name='conv2_3_3')(conv2_3_3_reduce)
        conv2_3_3 = LocalResponseNormalization(alpha=0.0001, beta=0.75)(conv2_3_3)
        pool2_3_3 = tf.keras.layers.MaxPooling2D(pool_size=3, strides=2, name='pool2_3_3_s2')(conv2_3_3)

        mavnet_3a_1_1 = tf.keras.layers.Conv2D(64, 1, activation='relu', padding='same', name='mavnet_3a_1_1')(pool2_3_3)
        mavnet_3a_3_3_reduce = tf.keras.layers.Conv2D(96,1, activation='relu', padding='same',  name='mavnet_3a_3_3_reduce')(pool2_3_3)
        mavnet_3a_3_3 = tf.keras.layers.Conv2D(128,kernel_size=3,  activation='relu', padding='same',  name = 'mavnet_3a_3_3')(mavnet_3a_3_3_reduce)
        mavnet_3a_pool = tf.keras.layers.MaxPooling2D(pool_size=3, strides=1, padding='same')(pool2_3_3)
        mavnet_3a_pool_1_1 = tf.keras.layers.Conv2D(32, kernel_size=1, activation='relu', padding='same', name='mavnet_3a_pool_1_1')(mavnet_3a_pool)
        mavnet_3a_output = tf.keras.layers.Concatenate(axis=3)([mavnet_3a_1_1, mavnet_3a_3_3, mavnet_3a_pool_1_1])

        mavnet_3b_1_1 = tf.keras.layers.Conv2D(128,kernel_size=1,activation='relu', padding='same', name= 'mavnet_3b_1_1' )(mavnet_3a_output)
        mavnet_3b_3_3_reduce = tf.keras.layers.Conv2D( 128, kernel_size=1, activation='relu', padding='same', name='mavnet_3b_3_3_reduce')(mavnet_3a_output)
        mavnet_3b_3_3 = tf.keras.layers.Conv2D( 192, kernel_size=3,  activation='relu', padding='same', name='mavnet_3b_3_3')(mavnet_3b_3_3_reduce)
        mavnet_3b_pool = tf.keras.layers.MaxPooling2D( pool_size=3, strides=1, padding='same', name='mavnet_3b_pool')(mavnet_3a_output)
        mavnet_3b_pool_1_1 = tf.keras.layers.Conv2D( 64, kernel_size=1,activation='relu', padding='same', name='mavnet_3b_pool_1_1')(mavnet_3b_pool)
        mavnet_3b_output = tf.keras.layers.Concatenate(axis=3,name='mavnet_3b_output')([mavnet_3b_1_1, mavnet_3b_3_3, mavnet_3b_pool_1_1])

        pool3_3_3 = tf.keras.layers.MaxPooling2D( pool_size=3, strides=2, padding='same', name='pool3_3_3')(mavnet_3b_output)

        mavnet_4a_1_1 = tf.keras.layers.Conv2D( 192, kernel_size=1, activation='relu', padding='same', name='mavnet_4a_1_1')(pool3_3_3)
        mavnet_4a_3_3_reduce = tf.keras.layers.Conv2D( 96, kernel_size=1, activation='relu', padding='same', name='mavnet_4a_3_3_reduce')(pool3_3_3)
        mavnet_4a_3_3 = tf.keras.layers.Conv2D( 208, kernel_size=3,  activation='relu', padding='same', name='mavnet_4a_3_3')(mavnet_4a_3_3_reduce)
        mavnet_4a_pool = tf.keras.layers.MaxPooling2D( pool_size=3, strides=1, padding='same', name='mavnet_4a_pool')(pool3_3_3)
        mavnet_4a_pool_1_1 = tf.keras.layers.Conv2D( 64, kernel_size=1, activation='relu', padding='same', name='mavnet_4a_pool_1_1')(mavnet_4a_pool)
        mavnet_4a_output = tf.keras.layers.Concatenate(axis=3, name='mavnet_4a_output')([mavnet_4a_1_1, mavnet_4a_3_3, mavnet_4a_pool_1_1])

        mavnet_4b_1_1 = tf.keras.layers.Conv2D( 160, kernel_size=1, activation='relu', padding='same', name='mavnet_4b_1_1')(mavnet_4a_output)
        mavnet_4b_3_3_reduce = tf.keras.layers.Conv2D( 112, kernel_size=1, activation='relu', padding='same', name='mavnet_4b_3_3_reduce')(mavnet_4a_output)
        mavnet_4b_3_3 = tf.keras.layers.Conv2D( 224, kernel_size=3, activation='relu', padding='same', name='mavnet_4b_3_3')(mavnet_4b_3_3_reduce)
        mavnet_4b_pool = tf.keras.layers.MaxPooling2D( pool_size=3, strides=1, padding='same', name='mavnet_4b_pool')(mavnet_4a_output)
        mavnet_4b_pool_1_1 = tf.keras.layers.Conv2D( 64, kernel_size=1, activation='relu', padding='same', name='mavnet_4b_pool_1_1')(mavnet_4b_pool)
        mavnet_4b_output = tf.keras.layers.Concatenate(axis=3, name='mavnet_4b_output')([mavnet_4b_1_1, mavnet_4b_3_3, mavnet_4b_pool_1_1])

        mavnet_4c_1_1 = tf.keras.layers.Conv2D( 128, kernel_size=1, activation='relu', padding='same', name='mavnet_4c_1_1')(mavnet_4b_output)
        mavnet_4c_3_3_reduce = tf.keras.layers.Conv2D( 128, kernel_size=1, activation='relu', padding='same', name='mavnet_4c_3_3_reduce')(mavnet_4b_output)
        mavnet_4c_3_3 = tf.keras.layers.Conv2D( 256,  kernel_size=3, activation='relu', padding='same', name='mavnet_4c_3_3')(mavnet_4c_3_3_reduce)
        mavnet_4c_pool = tf.keras.layers.MaxPooling2D( pool_size=3, strides=1, padding='same')(mavnet_4b_output)
        mavnet_4c_pool_1_1 = tf.keras.layers.Conv2D( 64, kernel_size=1, activation='relu', padding='same', name='mavnet_4c_pool_1_1')(mavnet_4c_pool)
        mavnet_4c_output = tf.keras.layers.Concatenate(axis=3,name='mavnet_4c_output')([mavnet_4c_1_1, mavnet_4c_3_3, mavnet_4c_pool_1_1])

        mavnet_4d_1_1 = tf.keras.layers.Conv2D( 112, kernel_size=1, activation='relu', padding='same', name='mavnet_4d_1_1')(mavnet_4c_output)
        mavnet_4d_3_3_reduce = tf.keras.layers.Conv2D( 144, kernel_size=1, activation='relu', padding='same', name='mavnet_4d_3_3_reduce')(mavnet_4c_output)
        mavnet_4d_3_3 = tf.keras.layers.Conv2D( 288, kernel_size=3, activation='relu', padding='same', name='mavnet_4d_3_3')(mavnet_4d_3_3_reduce)
        mavnet_4d_pool = tf.keras.layers.MaxPooling2D( pool_size=3, strides=1, padding='same', name='mavnet_4d_pool')(mavnet_4c_output)
        mavnet_4d_pool_1_1 = tf.keras.layers.Conv2D( 64, kernel_size=1, activation='relu', padding='same', name='mavnet_4d_pool_1_1')(mavnet_4d_pool)
        mavnet_4d_output = tf.keras.layers.Concatenate(axis=3, name='mavnet_4d_output')([mavnet_4d_1_1, mavnet_4d_3_3, mavnet_4d_pool_1_1])

        mavnet_4e_1_1 = tf.keras.layers.Conv2D( 256, kernel_size=1, activation='relu', padding='same', name='mavnet_4e_1_1')(mavnet_4d_output)
        mavnet_4e_3_3_reduce = tf.keras.layers.Conv2D( 160, kernel_size=1, activation='relu', padding='same', name='mavnet_4e_3_3_reduce')(mavnet_4d_output)
        mavnet_4e_3_3 = tf.keras.layers.Conv2D( 320, kernel_size=3, activation='relu', padding='same', name='mavnet_4e_3_3')(mavnet_4e_3_3_reduce)
        mavnet_4e_pool = tf.keras.layers.MaxPooling2D( pool_size=3, strides=1, padding='same', name='mavnet_4e_pool')(mavnet_4d_output)
        mavnet_4e_pool_1_1 = tf.keras.layers.Conv2D( 128, kernel_size=1, activation='relu', padding='same', name='mavnet_4e_pool_1_1')(mavnet_4e_pool)
        mavnet_4e_output = tf.keras.layers.Concatenate(axis=3)([mavnet_4e_1_1, mavnet_4e_3_3,mavnet_4e_pool_1_1])

        pool4_3_3 = tf.keras.layers.MaxPooling2D(pool_size=3, strides=2, padding='same', name='pool_3_3')(mavnet_4e_output)

        mavnet_5a_1_1 = tf.keras.layers.Conv2D( 256, kernel_size=1, activation='relu', padding='same', name='mavnet_5a_1_1')(pool4_3_3)
        mavnet_5a_3_3_reduce = tf.keras.layers.Conv2D( 160, kernel_size=1, activation='relu', padding='same', name='mavnet_5a_3_3_reduce')(pool4_3_3)
        mavnet_5a_3_3 = tf.keras.layers.Conv2D( 320, kernel_size=3, activation='relu', padding='same', name='mavnet_5a_3_3')(mavnet_5a_3_3_reduce)
        mavnet_5a_pool = tf.keras.layers.MaxPooling2D( pool_size=3, strides=1, padding='same', name='mavnet_5a_pool')(pool4_3_3)
        mavnet_5a_pool_1_1 = tf.keras.layers.Conv2D( 128, kernel_size=1,activation='relu', padding='same', name='mavnet_5a_pool_1_1')(mavnet_5a_pool)
        mavnet_5a_output = tf.keras.layers.Concatenate(axis=3)([mavnet_5a_1_1, mavnet_5a_3_3, mavnet_5a_pool_1_1])

        mavnet_5b_1_1 = tf.keras.layers.Conv2D( 384, kernel_size=1,activation='relu', padding='same', name='mavnet_5b_1_1')(mavnet_5a_output)
        mavnet_5b_3_3_reduce = tf.keras.layers.Conv2D( 192, kernel_size=1, activation='relu', padding='same', name='mavnet_5b_3_3_reduce')(mavnet_5a_output)
        mavnet_5b_3_3 = tf.keras.layers.Conv2D( 384,  kernel_size=3,activation='relu', padding='same', name='mavnet_5b_3_3')(mavnet_5b_3_3_reduce)
        mavnet_5b_pool = tf.keras.layers.MaxPooling2D( pool_size=3, strides=1, padding='same', name='mavnet_5b_pool')(mavnet_5a_output)
        mavnet_5b_pool_1_1 = tf.keras.layers.Conv2D( 128, kernel_size=1, activation='relu', padding='same', name='mavnet_5b_pool_1_1')(mavnet_5b_pool)
        mavnet_5b_output = tf.keras.layers.Concatenate(axis=3)([mavnet_5b_1_1, mavnet_5b_3_3, mavnet_5b_pool_1_1])

        pool5_7_7 = tf.keras.layers.AveragePooling2D(pool_size=7, strides=1, padding='same')(mavnet_5b_output)
        pool5_7_7 = tf.keras.layers.Dropout(0.4)(pool5_7_7)
        pool5_7_7_flatten = tf.keras.layers.Flatten()(pool5_7_7)
        output = tf.keras.layers.Dense(num_classes, activation='softmax')(pool5_7_7_flatten)

        model = tf.keras.Model(img_input, output)
        model.compile(optimizer=tf.keras.optimizers.SGD(0.0001), loss='categorical_crossentropy', metrics=['accuracy'])

        return model

    model = MavNet((100, 100, 1), num_classes)

    # Here, you can put input data of your own
    model.fit(data)

In addition, according to #37726, you can refer to a simple example in which gradient computation after `tf.concat` does not work : https://colab.research.google.com/drive/1dkCcL5jfBmo47EsvmhNumjIkCGIdeFd5

**Comment by joonjeon** [comment]
So... this is an issue with Keras, right?

If I understood your response correctly,
for all Tensorflow-"native" functions (i.e. the Tensorflow functions that are not part of Keras) for constructing neural network layers, almost of them including `tf.concat()` can have their gradients computed when properly put within `tf.GradientTape()` context, is that right?

If so, then I suppose I can close this issue for a while and mention it to the Keras Team as you have mentioned.

Thanks.
